### ISSUE TITLE
[HOLD for payment 2023-08-21] [HOLD for payment 2023-08-17] [$1000] mweb - Chat - Conversation freezes when sending a large PDF file

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#19918]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the testing link staging.new.expensify.com
2. Open any chat.
3. Send la large PDF in the conversation.

## Expected Result:
Verify the PDF preview is displayed and scrolling.

## Actual Result:
Conversation freezes when sending a large PDF file.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.21.1

**Reproducible in staging?:** yes

**Reproducible in production?:** yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**
https://expensify.testrail.io/index.php?/tests/view/3481149

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/e4c70442-4925-43c3-b7eb-3b0ba199790c

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01742a99d4a9530002</li>
        <li>Upwork Job ID: 1666564922927919104</li>
        <li>Last Price Increase: 2023-06-07</li>
    </ul>
</details>

**Comment by puneetlath** [comment]
@kbecciv can you share the PDF you used?

**Comment by kbecciv** [comment]
@puneetlath Please attached

[LARGE PDF FILE.pdf](https://github.com/Expensify/App/files/11639141/LARGE.PDF.FILE.pdf)

**Comment by puneetlath** [comment]
Added external label to see what contributors think.

**Comment by mountiny** [comment]
Thanks @rezkiy37, please remember to first post a proposal here for @0xmiroslav to review and discuss before you create a PR, thanks!

**Comment by puneetlath** [comment]
@0xmiroslav I'll be OOO this week, so if you need Internal help moving this forward during that time please ask in #contributor-plus. Thanks!

**Comment by rezkiy37** [comment]
Daily updates:

1. This is only a **Safari** issue. 
2. Looks like, after the updating [react-pdf](https://www.npmjs.com/package/react-pdf/v/6.2.2) to the last stable _6th_ version (_6.2.2_), fixes the issue on **MacOS Safari**.
3. Unfortunately, **IOS Safari** requires more steps to fix. [**IOS Safari** simply cannot draw large canvas elements. The limit is set at 16.777.216 pixels.](https://github.com/wojtekmaj/react-pdf/issues/1149#). Since the library renders _pdf_ as canvas, it means there is a default limitation of pages. Once the limit has been exhausted it draws empty pages without any content. 
4. There is [a solution](https://github.com/wojtekmaj/react-pdf/issues/1149#issuecomment-1323541137) how to open 140+ pages on **IOS Safari**, we could set [`devicePixelRatio`](https://github.com/wojtekmaj/react-pdf#props-1) manually. This reduces the quality of _pdf_ files a bit, things a little blurry. However, the files are still readable.
5. I am going to post a full proposal tomorrow.

**Comment by mountiny** [comment]
> There is https://github.com/wojtekmaj/react-pdf/issues/1149#issuecomment-1323541137 how to open 140+ pages on IOS Safari, we could set [devicePixelRatio](https://github.com/wojtekmaj/react-pdf#props-1) manually. This reduces the quality of pdf files a bit, things a little blurry. However, the files are still readable.

What other apps do in this scenario? I feel like if the pdf is this big, we should default to downloading it to Files or some management tool @0xmiroslav what do you think?

**Comment by rezkiy37** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The problem happens only in **Safari**. When the app tries to preview large _pdf_ files via `react-pdf` version [5.7.2](https://www.npmjs.com/package/react-pdf/v/5.7.2). It does not render content of the files.

### What is the root cause of that problem?
Looks like the current version (_5.7.2_) of `react-pdf` does not support the latest **Safari** version.
Also, **IOS Safari** has [a default limitation of canvas size](https://pqina.nl/blog/canvas-area-exceeds-the-maximum-limit/). Since, _pdf_ files are being rendered as canvas, `react-pdf` cannot render all pages.

### What changes do you think we should make in order to solve the problem?
Firstly, I would like to upgrade `react-pdf` to the last stable version - [6.2.2](https://www.npmjs.com/package/react-pdf/v/6.2.2). It supports the last **Safari** version. This upgrade requires following [the installation guide](https://github.com/wojtekmaj/react-pdf/tree/v6.2.2#configure-pdfjs-worker). Fortunately, this upgrade fixes **MacOS Safari** without any addition steps.
The package already has _7th_ version, but it was introduced about 1 month ago. I do not think that it is a stable enough.

Secondly, when users attach a _pdf_ file, the app can preview it, but should reduce the quality by setting [devicePixelRatio](https://github.com/wojtekmaj/react-pdf/issues/1149#issuecomment-1323541137) property manually to `Page` in `src/components/PDFView/index.js`.
No way, how to understand how many pages **IOS Safari** renders on different devices. A real device renders 60, but simulator 32. Based on this, I would set the lowest value - `1`, then it renders all pages, at least 144 everywhere.
There is a function `isMobileSafari`, I would use it to apply the property only for **IOS Safari**.
Actually, the pdf file is still readable.

![Screenshot 2023-06-13 at 16 26 04](https://github.com/Expensify/App/assets/57314631/ce097ed7-57a3-4193-9b2a-8076a0b8197e)


### What alternative solutions did you explore?
There are 2 alternative solutions:
1. When users want to check a _pdf_ file that was already uploaded, the app could open a new tab with this file itself. The app delegates preview work to the browser itself. This way no quality loss and no implicit downloads. In `src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js` I can improve `onPress` handler of `PressableWithoutFocus`. When both `isMobileSafari` and `Str.isPDF(source)` return `true` , the app executes `Linking.openURL(source)` to open a new tab with the file.

https://github.com/Expensify/App/assets/57314631/86b4c860-c910-49ae-9cc6-325c15ff8aab

2. When users open an uploaded file, the app could download it immediately. The app delegates preview work to **Files** of **IOS**.

**Comment by mountiny** [comment]
@0xmiroslav thoughts about the proposal?

**Comment by mountiny** [comment]
@rezkiy37 What do you think of the proposal from @allroundexperts seems reasonable to me.

**Comment by mountiny** [comment]
cc @0xmiroslav too

**Comment by rezkiy37** [comment]
> @rezkiy37 What do you think of the proposal from @allroundexperts seems reasonable to me.

I agree, it makes sense. I am trying to test this solution now. Let you know soon 🙂

**Comment by rezkiy37** [comment]
> After #20817, this issue will be converted to a crash. The PR is upgrading the react-pdf to v6.x. I think the package has been more optimized to lazy load the pdf pages thus the app does not freeze or crash immediately. It happens after the user scrolls a few pages.

I run the app with _6.2.2_, this upgrade fixes bug on **MacOS Safari**, but not on **IOS Safari**. Do you have some other insights?

**Comment by rezkiy37** [comment]
> I haven't dig this up but I think the best solution will be to fix it upstream.

I would like to have upgraded library first, then resolve this issue like a follow up. However, anyway, the bug exists on **IOS Safari**.

**Comment by rezkiy37** [comment]
#### Daily updates:
1. I've adapted successfully [this proposal](https://github.com/Expensify/App/issues/19918#issuecomment-1595348123) from @allroundexperts to our `PDFView` (class component). It works! (_check videos below_)
2. It makes sense to use [react-window](https://www.npmjs.com/package/react-window) to render pages of any _pdf_ document like a virtualised list on **web**.
3. Will post a full proposal tomorrow

#### Large _PDF_ file preview on **IOS Safari**
https://github.com/Expensify/App/assets/57314631/2a0690a4-3d21-40ba-b308-495800cb8d68

#### Large _PDF_ file preview on **MacOS Safari**
https://github.com/Expensify/App/assets/57314631/3e85fdcb-ebf5-482a-b866-85b989882538Daily updates:

**Comment by rezkiy37** [comment]
### Proposal

### Please re-state the problem that we are trying to solve in this issue.
The problem happens only in **Safari**. When the app tries to preview large _pdf_ files via `react-pdf` version _5.7.2_. It does not render content of the files.

### What is the root cause of that problem?
Looks like the current version (_5.7.2_) of `react-pdf` does not support the latest **Safari** version.
Also, **IOS Safari** has [a default limitation of canvas size](https://pqina.nl/blog/canvas-area-exceeds-the-maximum-limit/). Since, _pdf_ files are being rendered as canvas, `react-pdf` cannot render all pages.

### What changes do you think we should make in order to solve the problem?
There is solution: get rendered pages partially. It is similar as **React Native** works with **FlatList** for the native platforms. Since, the bug is on web (**Safari**), we do not have some similar virtualized list out of the box. Fortunately, [react-window](https://github.com/bvaughn/react-window) can help us to get desired behaviour.

1. Install `react-window`.

**Note**: All next steps (changes) should be applied only for the `PDFView` component.

2. Add 1 more value to **state** of the component: `pageViewports` (_array_). This value keeps an array of each page instance. It allows to get a page viewport to calculate an actual page height (check _5th_ step). Also, it helps to understand should or shouldn’t the component render the list of pages (check _7th_ step).

3. Implement `calculatePageWidth` method. The component already has this calculation. However, it is located in `render` method currently. By this proposal and next steps, we need share the value across other methods. Just copy all current calculations into the method.

4. Implement `renderPage` method. It is being used like a callback to render children of `VariableSizeList`. The method receives `index` and `style` (_specific styles with a position for each page_) like arguments. For `Page` I would use `pageIndex` instead of `pageNumber`. These props are pretty similar, but `pageIndex` has higher priority.
```js
renderPage({index, style}) {
  const pageWidth = this.calculatePageWidth()

  return (
      <View style={style}>
          <Page
              key={`page_${index}`}
              width={pageWidth}
              pageIndex={index}
              // This needs to be empty to avoid multiple loading texts which show per page and look ugly
              // See https://github.com/Expensify/App/issues/14358 for more details
              loading=""
          />
      </View>
  );
}
```

5. Implement `getPageHeight` method. Because, pages height is not a static value, the app should calculate it dynamically as well. It calculates an actual height of each page based on a scale.
```js
getPageHeight(pageIndex) {
  if (this.state.pageViewports.length === 0) {
      throw new Error('getPageHeight() called too early');
  }

  const pageViewport = this.state.pageViewports[pageIndex];
  const scale = this.calculatePageWidth() / pageViewport.width;
  const actualHeight = pageViewport.height * scale;
 
  return actualHeight;
}
```

6. Extend `onDocumentLoadSuccess` method. It should build an array of page instances - `pageViewports`. Also, it sets all previous values.
```js
onDocumentLoadSuccess(pdf) {
  const numPages = pdf.numPages;

  Promise.all(
      _.times(numPages, (index) => {
          const pageNumber = index + 1;

          return pdf.getPage(pageNumber).then((page) => page.getViewport({scale: 1}));
      }),
  ).then((pageViewports) => {
      this.setState({
          pageViewports,
          numPages,
          shouldRequestPassword: false,
          isPasswordInvalid: false,
      });
  });
}
```

7. Replace a block where the component renders pages. I am going to use `VariableSizeList` because a height of pages should be calculated dynamically.
```js
import {VariableSizeList as List} from 'react-window';

{
    this.state.pageViewports.length > 0 && (
        <List
            width={pageWidth}
            height={this.props.windowHeight}
            estimatedItemSize={this.getPageHeight(0)}
            itemCount={this.state.numPages}
            itemSize={this.getPageHeight}
        >
            {this.renderPage}
        </List>
    );
}
```

8. There is already #20817 where guys updated a version of `react-pdf`. I would like to have upgraded library first, then merge this proposal like a follow up. I confirm that this upgrade fixes the bug on **MacOS Safari**, but **not** on **IOS Safari**.

**Comment by rezkiy37** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/19918#issuecomment-1600603113)

**Comment by puneetlath** [comment]
@0xmiroslav what do you think of the proposal?

**Comment by rezkiy37** [comment]
Hi, guys @puneetlath, @0xmiroslav.
Any feedback on the [proposal](https://github.com/Expensify/App/issues/19918#issuecomment-1600603113) above?

**Comment by puneetlath** [comment]
@0xmiroslav what do you think?

**Comment by puneetlath** [comment]
@0xmiroslav friendly bump on reviewing the proposal. We're past 4 weeks now, so it'd be great to take it to the finish line asap.

**Comment by rezkiy37** [comment]
> Sorry I will update tomorrow

Hi! 
Any news?

**Comment by rezkiy37** [comment]
@0xmiroslav Just a friendly reminder 🙂

**Comment by rezkiy37** [comment]
> > I mean that he would not implement such fix under the hood of his package.
> 
> Can you link that comment from author?

Yes, I can.
1. https://github.com/wojtekmaj/react-pdf/issues/94#issuecomment-1308411320
> "without performance optimization" is the key here. React-PDF is NOT a PDF viewer - it is only a tool to build one. If you want to browse 100 page PDFs, you need to take similar precautions as if you were trying to open 100 images at once, or 100 videos, or whatever. You wouldn't open them all at once, would you?
2. https://github.com/wojtekmaj/react-pdf/issues/94#issuecomment-1308463549

**Comment by rezkiy37** [comment]
@0xmiroslav What do you think?

**Comment by rezkiy37** [comment]
@0xmiroslav Hi! Just a friendly reminder 🙂

**Comment by rezkiy37** [comment]
> Sounds like we need to fork our own repo (or patch?) if using react-window approach.
> 
> Also open for others' thoughts on [#19918 (comment)](https://github.com/Expensify/App/issues/19918#issuecomment-1623680240) which @mountiny recommended.
> 
> cc: @puneetlath (I see you're assigned as BZ here but also can take engineer role 🙂)

Based on author recommendation, I think we should add own wrapper (`react-window`) instead of adding/modifying something under the hood.

**Comment by rezkiy37** [comment]
> I thought it was helpful to community if our forked package can be widely used for performance improved PDF viewer.
> But yes, direct change in E/App is also fine to me.

Fine. I am implementing the proposal then. Thank you!

**Comment by puneetlath** [comment]
The proposal sounds good to me too. 

I think if you wanted to fork the package as something for Callstack or anyone else here (@rezkiy37 @allroundexperts etc) to own and maintain, that would be cool. Otherwise, directly updating our own codebase to use this also seems fine.

**Comment by puneetlath** [comment]
Yeah I think that's a cool way to do it if: 
1. We think it is an equally valid solution the problem
2. It will benefit others outside our codebase
3. There is someone who wants to create/maintain the package

**Comment by rezkiy37** [comment]
Hi, guys (@0xmiroslav, @puneetlath, @allroundexperts)!
So, if we are going create own package and apply it for this bug. Could you please clarify: how do we create own package step by step?

Also, I am considering the next variant:

1. Implement changes directly, updating our own codebase. It allows to test everything and make sure that it works as expected. BTW, it is easier to migrate to own package. 
2. Merge a PR and close this issue (_bug_).
3. Create a follow up issue. It should specialize only on the creating own package, integrating to our project, publishing. 

I mean, meanwhile I work on this issue to get it done, we create a follow up issue and work there together. 
What do you think?

**Comment by puneetlath** [comment]
That sounds fine to me. What do you think @0xmiroslav @allroundexperts?

**Comment by puneetlath** [comment]
Great, let's do it.

**Comment by rezkiy37** [comment]
@puneetlath, @0xmiroslav, @allroundexperts, I just opened an issue - https://github.com/Expensify/App/issues/22998, that we agreed. Please take a look at description. Also, I still need clarifications about how we create own package step by step. Please post a comment in the new issue. Feel free to comment the new issue.

**Comment by rezkiy37** [comment]
Work in progress here - https://github.com/Expensify/App/pull/24308.

**Comment by rezkiy37** [comment]
The PR (https://github.com/Expensify/App/pull/24308) is ready for review.

**Comment by rezkiy37** [comment]
Work in progress.

**Comment by puneetlath** [comment]
Perhaps we can split it 50/50. Does that seem fair to you two?

**Comment by situchan** [comment]
Fair to me

**Comment by puneetlath** [comment]
@0xmiroslav @situchan sent you both offers: 

https://www.upwork.com/nx/wm/offer/26515094
https://www.upwork.com/nx/wm/offer/26515111

**Comment by puneetlath** [comment]
Paid @situchan. 

@0xmiroslav bump on accepting the contract.

**Comment by puneetlath** [comment]
@0xmiroslav is tracking the payment he is due separately, so I'm going to go ahead and close his one out. Thanks everyone!
